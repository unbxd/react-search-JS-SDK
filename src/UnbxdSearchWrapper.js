import React, { Component } from 'react';
import PropTypes from 'prop-types';
import UnbxdSearch from '@unbxd-ui/unbxd-search-core';

import { AppContextProvider } from './common/context';
import {
    searchConfigurations as defaultSearchConfigurations,
    productTypes,
    searchStatus,
    paginationTypes
} from './config';
import {
    setProductConfiguration,
    setSearchBoxConfiguration,
    setSpellCheckConfiguration,
    setPageSizeConfiguration,
    setSortConfiguration,
    setMultilevelFacetsConfiguration,
    setRangeFacetsConfiguration,
    setFacetsActionConfiguration,
    setViewTypeConfiguration,
    manageTextFacets,
    manageRangeFacets,
    unbxdCallBack,
    getActiveFacets,
    handleViewTypeClick,
    getUpdatedResults,
    resetSearch,
    getAnalytics
} from './utils';
import { cloneElement } from './common/utils';
import '../public/css/core/index.scss';
import { viewTypes } from './config/constants';

/**
 * Component to initialize Unbxd Search. UnbxdSearchWrapper also acts as a root component for modules such as Products, Pagination and facets.
 */
const initialUnbxdState = {
    viewType: 'GRID',
    paginationType: paginationTypes.FIXED_PAGINATION,
    enableApplyFilters: false,
    selectedTextFacets: { add: {}, remove: {}, list: {} },
    selectedRangeFacets: { add: {}, remove: {}, list: {} },
    applyMultiple: false,
    pageSize: 10,
    sort: ''
};

class UnbxdSearchWrapper extends Component {
    constructor(props) {
        super(props);
        const {
            siteKey,
            apiKey,
            searchConfigurations,
            getCategoryId,
            setCategoryId,
            productType,
            priceUnit
        } = this.props;
        console.log('test')
        this.unbxdCallBack = unbxdCallBack.bind(this);
        this.setPageSizeConfiguration = setPageSizeConfiguration.bind(this);
        this.setSortConfiguration = setSortConfiguration.bind(this);
        this.setMultilevelFacetsConfiguration = setMultilevelFacetsConfiguration.bind(
            this
        );
        this.setRangeFacetsConfiguration = setRangeFacetsConfiguration.bind(
            this
        );
        this.setFacetsActionConfiguration = setFacetsActionConfiguration.bind(
            this
        );
        this.setProductConfiguration = setProductConfiguration.bind(this);
        this.setSearchBoxConfiguration = setSearchBoxConfiguration.bind(this);
        this.setSpellCheckConfiguration = setSpellCheckConfiguration.bind(this);
        this.setViewTypeConfiguration = setViewTypeConfiguration.bind(this);
        this.manageTextFacets = manageTextFacets.bind(this);
        this.manageRangeFacets = manageRangeFacets.bind(this);
        this.handleViewTypeClick = handleViewTypeClick.bind(this);
        this.getUpdatedResults = getUpdatedResults.bind(this);
        this.resetSearch = resetSearch.bind(this);
        this.getAnalytics = getAnalytics.bind(this);

        this.state = {
            unbxdCore: new UnbxdSearch({
                ...defaultSearchConfigurations,
                ...searchConfigurations,
                siteKey,
                apiKey,
                onEvent: this.unbxdCallBack,
                getCategoryId,
                setCategoryId
            }),
            productType,
            categoryId: '',
            unbxdState: initialUnbxdState,
            unbxdCoreStatus: searchStatus.READY,
            enableUnbxdAnalytics: searchConfigurations.enableUnbxdAnalytics,
            helpers: {
                setProductConfiguration: this.setProductConfiguration,
                setSearchBoxConfiguration: this.setSearchBoxConfiguration,
                setPageSizeConfiguration: this.setPageSizeConfiguration,
                setSortConfiguration: this.setSortConfiguration,
                setMultilevelFacetsConfiguration: this
                    .setMultilevelFacetsConfiguration,
                setRangeFacetsConfiguration: this.setRangeFacetsConfiguration,
                setFacetsActionConfiguration: this.setFacetsActionConfiguration,
                setSpellCheckConfiguration: this.setSpellCheckConfiguration,
                setViewTypeConfiguration: this.setViewTypeConfiguration,
                manageTextFacets: this.manageTextFacets,
                manageRangeFacets: this.manageRangeFacets,
                handleViewTypeClick: this.handleViewTypeClick,
                getUpdatedResults: this.getUpdatedResults,
                resetSearch: this.resetSearch,
                getAnalytics: this.getAnalytics,
                getActiveFacets
            },
            priceUnit
        };

        this.initialResultLoad = true;
    }

    componentDidMount() {
        const { unbxdCore } = this.state;
        const { trackCategory } = this.getAnalytics();
        const categoryId =
            typeof unbxdCore.options.getCategoryId === 'function' &&
            unbxdCore.options.getCategoryId();

        const viewType =
            this.state.unbxdCore.getQueryParams()['viewType'] || viewTypes.GRID;
        unbxdCore.options.extraParams['viewType'] = viewType;

        if (unbxdCore.options.applyMultipleFilters) {
            this.setState((currentState) => {
                return {
                    ...currentState,
                    unbxdState: {
                        ...currentState.unbxdState,
                        enableApplyFilters: true
                    }
                };
            });
        }

        if (
            categoryId &&
            typeof categoryId === 'string' &&
            categoryId.length > 0
        ) {
            this.setState((currentState) => {
                return {
                    ...currentState,
                    categoryId,
                    productType: productTypes.CATEGORY
                };
            });
            unbxdCore.options.productType = productTypes.CATEGORY;
            unbxdCore.getResults();
            trackCategory(window.UnbxdAnalyticsConf);
        }

        if (unbxdCore.options.hashMode) {
            window.onhashchange = unbxdCore.onLocationChange.bind(unbxdCore);
        }
    }

    componentDidUpdate(prevProps) {
        const { refreshId } = prevProps;
        const { productType } = this.props;
        const { unbxdCore, categoryId } = this.state;
        const { trackCategory } = this.getAnalytics();

        const urlParams = unbxdCore.getQueryParams();
        const getResults = unbxdCore.getResults.bind(unbxdCore);
        const renderFromUrl = unbxdCore.renderFromUrl.bind(unbxdCore);
        const currentCategoryId =
            typeof unbxdCore.options.getCategoryId === 'function' &&
            unbxdCore.options.getCategoryId();

        if (
            categoryId !== currentCategoryId &&
            productType === productTypes.CATEGORY &&
            typeof currentCategoryId === 'string' &&
            currentCategoryId.length > 0
        ) {
            this.setState((currentState) => {
                return {
                    ...currentState,
                    categoryId: currentCategoryId,
                    productType: productTypes.CATEGORY
                };
            });
            unbxdCore.options.productType = productTypes.CATEGORY;
            if (categoryId.length === 0 && Object.keys(urlParams).length) {
                renderFromUrl();
            } else {
                this.resetSearch();
                getResults();
            }
            trackCategory(window.UnbxdAnalyticsConf);
        } else if (
            unbxdCore.getResponseObj() === null &&
            urlParams[unbxdCore.options.searchQueryParam]
        ) {
            renderFromUrl();
        } else if (refreshId !== this.props.refreshId) {
            this.resetSearch();
            getResults();
        }
    }

    getProps() {
        const {
            unbxdCore,
            unbxdCoreStatus,
            unbxdState,
            helpers,
            priceUnit,
            productType
        } = this.state;
        return {
            unbxdCore,
            unbxdCoreStatus,
            unbxdState,
            helpers,
            priceUnit,
            productType
        };
    }

    static getDerivedStateFromProps(props, state) {
        if (props.productType !== state.productType) {
            return { productType: props.productType };
        }

        return null;
    }

    render() {
        const { unbxdCoreStatus } = this.state;
        const { loaderComponent, errorComponent } = this.props;

        return (
            <>
                {unbxdCoreStatus === searchStatus.LOADING &&
                    loaderComponent &&
                    cloneElement(loaderComponent)}

                {unbxdCoreStatus === searchStatus.ERROR &&
                    errorComponent &&
                    cloneElement(errorComponent)}

                <AppContextProvider value={this.getProps()}>
                    {this.props.children}
                </AppContextProvider>
            </>
        );
    }
}

UnbxdSearchWrapper.defaultProps = {
    productType: 'SEARCH',
    priceUnit: '$',
    searchConfigurations: {}
};

UnbxdSearchWrapper.propTypes = {
    /**
     * Site key of the site.
     */
    siteKey: PropTypes.string.isRequired,
    /**
     * API key of the site.
     */
    apiKey: PropTypes.string.isRequired,
    /**
     * Custom function to return the Category Id.
     */
    getCategoryId: PropTypes.func,
    /**
     * Custom function to set the Category Id.
     */
    setCategoryId: PropTypes.func,
    /**
     * Product type of UnbxdSearchWrapper.
     */
    productType: PropTypes.string,
    /**
     * Price unit.
     */
    priceUnit: PropTypes.string,
    /**
     * custom loader component.
     */
    loaderComponent: PropTypes.element,
    /**
     * custom error component instance.
     */
    errorComponent: PropTypes.element,
    /**
     * id to trigger a refresh.
     */
    refreshId: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    /**
     * search configurations object.
     */
    searchConfigurations: PropTypes.object,
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ])
};

export default UnbxdSearchWrapper;
