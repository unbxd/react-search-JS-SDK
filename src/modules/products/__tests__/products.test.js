import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react'

import React from 'react'
import Products from '../index';
import UnbxdSearchWrapper from '../../../UnbxdSearchWrapper';
import SearchBox from '../../searchBox';

test('Match Snapshot for products', () => {
    window.fetch = jest.fn(() => Promise.resolve());
    const attributesMap = {
        productName: 'title',
        uniqueId: 'uniqueId',
        imageUrl: 'imageUrl',
        price: 'min_cheapest_default_price',
        productUrl: 'productUrl',
    };
    const tree = renderer.create((
        <UnbxdSearchWrapper
            siteKey="wildearthclone-neto-com-au808941566310465"
            apiKey="e6959ae0b643d51b565dc3e01bf41ec1"
        >
            <Products attributesMap={attributesMap} />
            <div>
                <SearchBox defaultSearch='shoes' />
            </div>
        </UnbxdSearchWrapper>
    )).toJSON();
    expect(tree).toMatchSnapshot();
});

test('No products found test', () => {
    window.fetch = jest.fn(() => Promise.resolve());
    const attributesMap = {
        productName: 'title',
        uniqueId: 'uniqueId',
        imageUrl: 'imageUrl',
        price: 'min_cheapest_default_price',
        productUrl: 'productUrl',
    };
    render(
        <UnbxdSearchWrapper
            siteKey="wildearthclone-neto-com-au808941566310465"
            apiKey="e6959ae0b643d51b565dc3e01bf41ec1"
        >
            <Products attributesMap={attributesMap} />
            <div>
                <SearchBox defaultSearch='shoes' />
            </div>
        </UnbxdSearchWrapper>
    );
    expect(screen.getByText('Sorry! No products found!')).toBeInTheDocument();
});
