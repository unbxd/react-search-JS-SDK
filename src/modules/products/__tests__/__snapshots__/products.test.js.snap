// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match Snapshot for products 1`] = `
Array [
  <div>
    <div
      className="UNX-products__container"
    >
      <div
        className="UNX-products__list -grid grid-cols-4"
      />
    </div>
  </div>,
  <div>
    <div
      className="UNX-searchbox__container"
    >
      <form
        onSubmit={[Function]}
      >
        <input
          className="UNX-searchbox__input"
          onChange={[Function]}
          placeholder=""
          value=""
        />
        
        <button
          className="UNX-searchbox__button"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
  </div>,
]
`;
